java-buildpack/bin/build

```diff
@@ -0,0 +1,56 @@
+#!/usr/bin/env bash
+
+echo "---> Java buildpack"
+
+set -eo pipefail
+
+env_dir=$1/env
+cache_dir=$2
+launch_dir=$3
+
+# Load user-provided build-time environment variables
+if compgen -G "$env_dir/*" > /dev/null; then
+  for var in "$env_dir"/*; do
+    declare "$(basename "$var")=$(<"$var")"
+  done
+fi
+
+echo "---> Installing JDK"
+
+if [[ ! -f $launch_dir/jdk.toml ]]; then
+  mkdir -p $launch_dir/jdk
+  jdk_url="https://cdn.azul.com/zulu/bin/zulu8.28.0.1-jdk8.0.163-linux_x64.tar.gz"
+  curl -sfL "$jdk_url" | tar pxz -C $launch_dir/jdk --strip-components=1
+  mkdir -p $launch_dir/jdk/.profile.d
+  cat << EOF > $launch_dir/jdk/.profile.d/jdk.sh
+export JAVA_HOME=$launch_dir/jdk
+export PATH=\$JAVA_HOME/bin:\$PATH
+export LD_LIBRARY_PATH="\$JAVA_HOME/jre/lib/amd64/server:\$LD_LIBRARY_PATH"
+EOF
+  echo "version = \"1.8.0_163\"" > $launch_dir/jdk.toml
+
+  # TODO put the jdk in the cache
+  # TODO write to $cache_dir/jdk/env
+fi
+source $launch_dir/jdk/.profile.d/jdk.sh
+
+echo "---> Installing Maven"
+
+mkdir -p $cache_dir/maven/.m2
+if [[ ! -f $cache_dir/maven.toml ]]; then
+  maven_url="https://apache.org/dist/maven/maven-3/3.5.4/binaries/apache-maven-3.5.4-bin.tar.gz"
+  curl -sfL "$maven_url" | tar pxz -C "$cache_dir/maven" --strip-components=1
+  echo "version = \"3.5.4\"" > $cache_dir/maven.toml
+fi
+export PATH=$PATH:$cache_dir/maven/bin
+
+echo "---> Running Maven"
+
+MAVEN_OPTS="${MAVEN_OPTS:-"-Xmx1024m -Duser.home=$cache_dir -Dmaven.repo.local=$cache_dir/.m2/repository"}"
```

Comment:
Traditionally we've [symlinked a cached directory to `$HOME/.m2`](https://github.com/cloudfoundry/cf-java-client/blob/master/ci/unit-test.sh#L5) because there's no combination of environment variables and system properties that will move all caches (say the `mvnw` download cache).  @sclevine would there be any problems doing that, knowing that the symlink would disappear when the container was disposed and that'd be fine?

Reply:
Should work great. `/home/pack` (`$HOME`) is intended for this purpose -- an ephemeral home directory that is not preserved.

Reply:
@nebhale in the Heroku Java buildpack we let it do whatever `/app` and then move it to the cache after the build runs. I'm just being lazy here (and was expecting this to change when i support `mvnw`).

@sclevine I'm not sure how I feel about relying on `$HOME`. I don't think we've made that a part of the API, and on Heroku `$HOME` is actually the `/app` dir. 

If it's ok with y'all, I'll merge this as is, so I have a baseline to iterate from.

Reply:
Interesting point about `$HOME`. CF is super weird with this: it's `/home/vcap` at build, but `/home/vcap/app` (symlinked to `/app`) at launch. I think it should be the same in both cases, and I think I have a slight preference for `/home/pack`, because it prevents auto-generated settings/cache from various CLI tools from persisting (ex. `~/.bundle`).

@nebhale any thoughts about this?

Reply:
@jkutner I'm totally okay with you adding any samples to this repo directly 😄

Reply:
I’d agree with *definitely* ditching the CF insanity, and also lean towards home being somewhere other than the app directory. My anecdotal experience tells me that there are enough packages that write to `$HOME` that it’s a vector for vulnerabilities. Choosing a “safer” location seems like a prudent safety measure, but I am sensitive to compatibility.